From 5dbe19b5f61b9a3fe2a18dd8e9ecbea74503544f Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Fri, 7 May 2021 17:33:03 +0800
Subject: [PATCH] support to generate iso out of yocto build

- Support adding files/dirs to specific dest dir

- Tweak initramfs install dir, append cr_workdir prefix

Upstream-Status: Inappropriate [oe specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 .../lib/wic/plugins/source/isoimage-isohybrid.py   | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/scripts/lib/wic/plugins/source/isoimage-isohybrid.py b/scripts/lib/wic/plugins/source/isoimage-isohybrid.py
index 21740ac7c6..3c6603ff06 100644
--- a/scripts/lib/wic/plugins/source/isoimage-isohybrid.py
+++ b/scripts/lib/wic/plugins/source/isoimage-isohybrid.py
@@ -226,7 +226,15 @@ class IsoImagePlugin(SourcePlugin):
             payload_dir = source_params['payload_dir']
 
             logger.debug("Payload directory: %s", payload_dir)
-            shutil.copytree(payload_dir, iso_dir, symlinks=True, dirs_exist_ok=True)
+
+            if source_params.get('dest_dir'):
+                dest_dir = os.path.join(iso_dir, source_params['dest_dir'])
+            else:
+                dest_dir = iso_dir
+
+            logger.debug("Dest directory: %s", dest_dir)
+
+            shutil.copytree(payload_dir, dest_dir, symlinks=True, dirs_exist_ok=True)
 
     @classmethod
     def do_prepare_partition(cls, part, source_params, creator, cr_workdir,
@@ -284,8 +292,8 @@ class IsoImagePlugin(SourcePlugin):
             if not os.path.isfile(initrd):
                 initrd = cls._build_initramfs_path(rootfs_dir, cr_workdir)
 
-        install_cmd = "install -m 0644 %s %s/initrd" % (initrd, isodir)
-        exec_cmd(install_cmd)
+        install_cmd = "install -m 0644 %s/%s %s/initrd" % (cr_workdir, initrd, isodir)
+        exec_cmd(install_cmd, as_shell=True)
 
         # Remove the temporary file created by _build_initramfs_path function
         if os.path.isfile("%s/initrd.cpio.gz" % cr_workdir):
-- 
2.27.0

